### DISCUSSION TITLE
Trajopt for trajectory interpolation

### DISCUSSION THREAD

**Discussion by johnwason** [tesseract-robotics/tesseract#804] [Q&A]
I have a set of waypoints that I need to interpolate into a set of dense waypoints with uniform time (ie every 4 ms), with jerk-limiting and ideally cartesian interpolation. How do I correctly formulate the problem to have trajopt do this interpolation?

A related question, can trajopt do something similar to ruckig with online updating? How do I formulate that problem?

**Comment by Levi-Armstrong** [comment]
The interpolation is done by the SimplePlanner which you can either use one of the existing with a modified profile or implement your own simple planner. Then this feeds into the trajopt planner. If you use the trajopt planner in the planning server it will first run the program through the simple planner and then through trajopt. 

> A related question, can trajopt do something similar to ruckig with online updating? How do I formulate that problem?

Yes, this was the main reason behind the development of trajopt_ifopt and the online example is a basic implementation. trajopt_ifopt is almost finished but does need some work before it can completely replace the current trajopt, but it may have what you need.

**Reply by johnwason** [reply]
How do I request a time optimal trajectory out of trajopt? The last time we used it, we had to use an output filter to get time stamps.

**Reply by johnwason** [reply]
We used the uniform_sample_filter when we used moveit in a previous project: https://github.com/ros-industrial/industrial_core/blob/melodic-devel/industrial_trajectory_filters/include/industrial_trajectory_filters/uniform_sample_filter.h . Have we pulled this type of filter in?

**Comment by Levi-Armstrong** [comment]
> How do I request a time optimal trajectory out of trajopt? The last time we used it, we had to use an output filter to get time stamps.

Using the default trajopt you still need to run it through a post filter which run time parameterization. This main reason for developing trajopt_ifopt was because it is very easy to introduce new parameters like time, but the current focus is to get to have the same capability as the original.

> Have we pulled this type of filter in?

There are several simple planners depending on what you need. There is one that is similar which leverages both joint space and cartesian distance threshold to do the interpolation.

**Reply by emfebert** [reply]
Are you referring to this file: https://github.com/tesseract-robotics/tesseract_planning/blob/master/tesseract_motion_planners/simple/src/simple_motion_planner.cpp ?
Where is it doing the time parametrization?
Do you have any examples for using the simple planner?
